RFC: rtio: SPI and a Fast Sensor

teburd opened the pull request:
Many changes in this RFC PR are very unlikely to make it upstream as-is or perhaps ever but as a practical demonstration for how RTIO could be used with a sensor, on a spi bus it does just that.

The sample app itself (samples/boards/tdk_robokit1) shows how delays may result in buffer underruns/overruns (depending on the perspective!), and how they may be resolved by having more buffers to feed the demand of the sensor. It would be interesting to also consider a throughput/latency comparison... lots of small buffers vs a few large ones.

It also lets you build with and without CONFIG_SPI_RTIO to show how thread stacks might be saved by using RTIO with SPI. Once #55178 is merged I'll rebase again and we can do probably even more compare/contrasts to get a sense of things.

This builds on top of #55382

Commit:
sensor: RTIO API support for FIFO streaming

Provides the necessary APIs to stream a sensors FIFO using RTIO

Adds an API call to enable explicitly reading the fifo from the sensor
at any time. The thinking is that while the sensors gpio trigger might
be one source of triggering a fifo read, external sources should be
possible.

Signed-off-by: Tom Burdick <thomas.burdick@intel.com>

tbursztyka commented on include/zephyr/drivers/spi.h:
```diff
@@ -878,6 +918,117 @@ __deprecated static inline int spi_write_async(const struct device *dev,
+}
+
+/**
+ * @brief Convert a transcieve from spi_buf_sets to a sequence of RTIO ops
```
s/transceive/transceive  

Typo seem to repeat in many place below.

teburd replied:
Thank you, yes I'm sure there's many other things in here that will need correction :-)

Commit:
icm42688: RTIO device for streaming

An experimental idea to enable streaming fifo data from the sensor using
RTIO with the existing sensor API. A sensor reader could easily be built on
top of the buffer to produce meaningful values.

There are many ways this can be built currently. Using either the existing
trigger work context (system workqueue or own thread) when SPI_RTIO=n or
using a sequence. With or without SPI_SAM_DMA (where only polling SPI
transfers are used).

This is a really nice setup to explore and experiment with RTIO, SPI, and
a really fast sensor.

Signed-off-by: Tom Burdick <thomas.burdick@intel.com>

Commit:
samples: boards: TDK Robokit1 Sensor Sample

Provides a sample demonstrating the fast sensor on a tdk robokit1 used
with and without RTIO.

With RTIO the sample may be built with SPI_RTIO enabled or disabled to
get a sense of the cost of a filler thread when SPI does not support RTIO
as well.

Signed-off-by: Tom Burdick <thomas.burdick@intel.com>

teburd commented:
Patch set is mostly cleaned up now and worth perusing for criticisms and such.

Commit:
samples: tdk_robokit1: Poll the magnetometer

While an asynchronous transfer is on going for the 6 axis imu, concurrently
read the magnetometer with its blocking fetch call.

In the future it'd be nice to have the magnetometer also work with RTIO
driven by a timer. In fact having all the sensors read on a timer would be
incredibly nice to showcase.

Signed-off-by: Tom Burdick <thomas.burdick@intel.com>

MaureenHelm commented on include/zephyr/rtio/rtio.h:
```diff
@@ -133,20 +134,34 @@ struct rtio_sqe {
+		struct {
+			void (*callback)(struct rtio *r, struct rtio_sqe *sqe, void *arg0);
+			void *arg0;
+		};
```
Is this change related to `rtio: Add tiny write op`?

MaureenHelm commented on include/zephyr/drivers/spi.h:
```diff
@@ -878,6 +892,47 @@ __deprecated static inline int spi_write_async(const struct device *dev,
+ *
+ * @param node DT_NODE
+ */
+#define SPI_IODEV_DEFINE(name, node_id, operation_, delay_)			\
+	const struct spi_dt_spec _spi_dt_spec_##name =				\
+		SPI_DT_SPEC_GET(node_id, operation_, delay_);			\
+	RTIO_IODEV_DEFINE(name, &spi_iodev_api, (void *)&_spi_dt_spec_##name)
```
`SPI_IODEV_DT_DEFINE` ?

MaureenHelm commented on include/zephyr/drivers/sensor.h:
```diff
@@ -593,6 +634,70 @@ static inline int z_impl_sensor_channel_get(const struct device *dev,
+	const struct sensor_driver_api *api =
+		(const struct sensor_driver_api *)dev->api;
+
+	return api->fifo_iodev(dev, iodev);
```
Might want to check if `fifo_iodev` is NULL

MaureenHelm commented on drivers/sensor/icm42688/Kconfig:
```diff
@@ -4,6 +4,8 @@
 #
 # SPDX-License-Identifier: Apache-2.0
 
+DT_COMPAT_ICM42688 := invensense,icm42688
```
Most (all?) of these have been removed. Is this a rebasing mistake?

teburd replied:
Almost certainly

MaureenHelm reviewed (commented):
some nitpicky comments, but overall this looks very cool!
